### PR TITLE
collections: posts as collection

### PR DESCRIPTION
**Pull request by parkr**
I guess I need to release Jekyll 3 tomorrow...

This is the last piece we were considering doing.

In an effort to not break sites, this is all backwards compatible liquid-wise. The only big difference is that `site.posts` is a `Collection` instead of an `Array`. But I feel more comfortable breaking plugins than I do breaking people's Liquid. Especially because Liquid doesn't have a means of erroring out on a bad dereference (see https://github.com/Shopify/liquid/issues/490 for more).

The cukes all pass. I'm going to work on the presentation now and leave this. But here it is.

/cc @benbalter who was looking to see this in 3.

### CODE REVIEW

**Comment by parkr** [comment]
cc @jekyll/core for thoughts. will merge now – please leave thoughts or submit follow-up PR's in the 3.0 milestone to fix things. it's not perfect, but at least it's something

File: lib/jekyll/configuration.rb [hunk 1]
```diff
@@ -275,6 +275,21 @@ def fix_common_issues
...
+
+      if config['collections'].is_a?(Array)
+        config['collections'] = Hash[config['collections'].map{|c| [c, {}]}]
+      end
+      config['collections']['posts'] ||= {}
+      config['collections']['posts']['output'] = true
+      config['collections']['posts']['permalink'] = style_to_permalink(config['permalink'])
+
+      config
+    end
```
Comment (benbalter) [inline lib/jekyll/configuration.rb]: Nicely implemented.

File: lib/jekyll/document.rb [hunk 2]
```diff
@@ -242,19 +293,54 @@ def read(opts = {})
       end
     end
 
+    def post_read
```
Comment (benbalter) [inline lib/jekyll/document.rb]: Does this mean that all collections can have a date now?

File: lib/jekyll/document.rb [hunk 3]
```diff
@@ -242,19 +293,54 @@ def read(opts = {})
...
           "content"       => content,
           "relative_path" => relative_path,
           "path"          => relative_path,
           "url"           => url,
-          "collection"    => collection.label
-        }
+          "collection"    => collection.label,
+          "next"          => next_doc,
+          "previous"      => previous_doc,
+          "id"            => id,
```
Comment (benbalter) [inline lib/jekyll/document.rb]: All documents having IDs is going to fix so many terrible workarounds I've created. :smile:

**Comment by benbalter** [comment]
Helz yes

**Comment by envygeeks** [comment]
Since @benbalter posted the Wonder Years guy is this gonna be Jekyll's Wonder Years?

**Comment by benbalter** [comment]
Technically, that was from FOX's new hit show [the Grinder](http://www.imdb.com/title/tt4481322/), so I say we have a 50/50 chance. :smile:

File: lib/jekyll/document.rb (hunk 2 above)
Comment (parkr) [inline lib/jekyll/document.rb]: Yes!

**Comment by parkr** [comment]
Thanks for taking the time to :eyes: this, @benbalter!

**Comment by envygeeks** [comment]
He will forever be Kevin Arnold to me. Always. :100:

File: lib/jekyll/document.rb (hunk 2 above)
Comment (benbalter) [inline lib/jekyll/document.rb]: Huh. :metal:

Comment (parkr) [inline lib/jekyll/document.rb]: `_decisions/2015-06-29-glossip-v-gross.md`

**Comment by alfredxing** [comment]
Woohoo! :beers: 

I'll take a good look at the code a bit later!

**Pull request merged**